### PR TITLE
Update pulse alignment context

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary

This PR updates output data of `AlignmentKind.to_dict` and ~deprecates `AlignFunc` context for qpy serialization~.

### Details and comments

Now output of `to_dict` is identical to the signature of the context class constructor, and we can serialize/deserialize the context via the output dictionary. `AlignFunc` is deprecated because it takes arbitrary callable to dynamically decide position of context instructions, and this cannot be serialized in qpy format. The impact of deprecation is expected to be small since this context is not used in tutorials. 

(New direcation)
`to_dict` is just replaced by `context_params` property which is a tuple of parameter values and alignment classes provide `is_serializable` attribute to tell qpy serializer if it can be serialized.

### CODE REVIEW

File: qiskit/pulse/builder.py [hunk 1]
```diff
@@ -1055,7 +1056,7 @@ def align_func(
 
     Examples:
 
-    .. jupyter-execute::
+    .. code-block:: python
```
Comment (eggerdj) [inline qiskit/pulse/builder.py:1059]: Why this change? Because of deprecation?

File: qiskit/pulse/transforms/alignments.py [hunk 2]
```diff
@@ -48,17 +48,15 @@ def align(self, schedule: Schedule) -> Schedule:
 
     def to_dict(self) -> Dict[str, Any]:
```
Comment (eggerdj) [inline qiskit/pulse/transforms/alignments.py:49]: This is a bit confusing. Should this not be an abstract method now?

File: qiskit/pulse/transforms/alignments.py [hunk 3]
```diff
@@ -283,11 +281,11 @@ def align(self, schedule: Schedule) -> Schedule:
...
         """Returns dictionary to represent this alignment."""
-        return {"alignment": self.__class__.__name__, "duration": self.duration}
+        return {"duration": self.duration}
 
 
 class AlignFunc(AlignmentKind):
-    """Allocate instructions at position specified by callback function.
+    """Deprecated. Allocate instructions at position specified by callback function.
```
Comment (eggerdj) [inline qiskit/pulse/transforms/alignments.py:288]: Is there a decorator we could use to indicate that `AlignFunc` is deprecated?

**Review by eggerdj — CHANGES_REQUESTED** [review]
I see why `AlignFunc` is annoying from a code perspective due to serialization. However, I also see that `AlignFunc` is useful from the perspective of DD. Since we can reproduce this with delay's it is probably okay to deprecate this alignment context. How would we recover this functionality without delays? Would we have a library of alignment contexts used for DD, e.g. where the function is essentially hard coded into the alignment context?

File: qiskit/pulse/transforms/alignments.py (hunk 2 above)
Comment (nkanazawa1989) [reply qiskit/pulse/transforms/alignments.py:64]: This is deprecated and now it implements `context_params` property for parameter assignment and context serialization. This is the same idea as `Instruction` implements `.operands` property.

File: qiskit/pulse/builder.py (hunk 1 above)
Comment (nkanazawa1989) [reply qiskit/pulse/builder.py]: This is no longer relevant but you cannot build document with jupyter-execute that executes deprecated code (because warning is handled as error).

**Comment by nkanazawa1989** [comment]
Thanks for the comment. Indeed I didn't think deprecation just for serialization is smart solution. Now alignments implement `is_serializable` class attribute that can be used for qpy serializer. If this property is `False` probably qpy serializer raises an error and terminate serialization. Another option is to serialize arbitrary function with Python AST, but I think this is overkill.

File: releasenotes/notes/deprecate-align-func-c710e62e01595f01.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+deprecations:
+  - |
+    Pusle alignment class method :meth:`~qiskit.pulse.transforms.alignments.AlignmentKind.to_dict`
+    has been deprecated and will be removed in a future release. This is replaced with
+    property :meth:`~qiskit.pulse.transforms.alignments.AlignmentKind.context_params`.
```
Comment (eggerdj) [inline releasenotes/notes/deprecate-align-func-c710e62e01595f01.yaml:6]:
```suggestion
    The pulse alignment class method :meth:`~qiskit.pulse.transforms.alignments.AlignmentKind.to_dict`
    has been deprecated and will be removed in a future release. This method is replaced with the
    property :meth:`~qiskit.pulse.transforms.alignments.AlignmentKind.context_params`.
```

**Review by eggerdj — CHANGES_REQUESTED** [review]
Minor suggestions looks good otherwise.

File: qiskit/pulse/transforms/alignments.py [hunk 5]
```diff
@@ -9,27 +9,36 @@
...
 from qiskit.pulse.exceptions import PulseError
 from qiskit.pulse.schedule import Schedule, ScheduleComponent
-from qiskit.pulse.utils import instruction_duration_validation
+from qiskit.pulse.utils import instruction_duration_validation, deprecated_functionality
 
 
 class AlignmentKind(abc.ABC):
     """An abstract class for schedule alignment."""
 
+    is_serializable = True
```
[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit/pulse/transforms/alignments.py:31]: Make sense

[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit/pulse/transforms/alignments.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7310/commits/bb1a8afe1a26a02ffce8e09bbfba15f817531dad

**Review by eggerdj — APPROVED** [review]
LGTM.

**Comment by nkanazawa1989** [comment]
Closed because this is a part of #7300
